File: neetcode all and misc questions/Heaps/longest_happy_string.py
class Solution:
    def longestDiverseString(self, a: int, b: int, c: int) -> str:
        """
        string cannot contain the substring with three of the same characters consecutively
        each character cannot exceed the amount listed for each parameter
        (count of char "a" cannot exceed the value in parameter "a")
        
        get the longest possible happy string
        
        In a similar way to the problem "Reorganize string",
        we can try a similar strategy where you start with the character
        that appears the most. If you reach a point where
        two characters (i.e "aa") have been added, you need to put
        pick the next most frequent character
        """
        import heapq
        maxHeap = []
        options = {"a": a, "b": b, "c": c}
        for key in options:
            if (options[key] != 0):
                heapq.heappush(maxHeap, (-options[key], key))
        cur = ""
        while (maxHeap):
            count, char = heapq.heappop(maxHeap)
            count = -1 * count
            # if we have two consecutive characters already, we cannot
            # add the third, so we have to pick to the next most frequently
            # occuring character instead
            if len(cur) >= 2 and cur[-1] == cur[-2] and cur[-1] == char:
                # if there is no "next" frequently occurring character,
                # break
                if (not maxHeap):
                    break
                count2, char2 = heapq.heappop(maxHeap)
                count2 = -1 * count2
                cur += char2
                count2 -= 1
                # if there are still characters left for the "next" frequently
                # occurring character, push back into the heap
                if count2 > 0:
                    heapq.heappush(maxHeap, (-1 * count2, char2))
            # if we did not reach the condition where we've added two characters in a row (or the max amount of characters we could've added for this character)
            else:
                count -= 1
                cur += char    
            # push back into the heap with the updated value for count
            if (count > 0):
                heapq.heappush(maxHeap, (-1 * count, char))
        return cur
            
            
            

File: node_js_ds_and_algos_implementations/heaps/longest_happy_string.js
class Solution {
    /**
     * @param {number} a
     * @param {number} b
     * @param {number} c
     * @return {string}
     */
    longestDiverseString(a, b, c) {
        /* 
        no aaa, bbb, ccc means that if you choose "a"
        you need a cooldown time of at least one to avoid picking more than two in a row

        Heap problem?
        we already have the counts of each character
        max heap, use the character with the most amount first to get the longest
        possible string while still respecting the cooldown time

        you should also maximize picking two characters if you can over just one
        */
        let maxHeap = new MaxPriorityQueue((x) => x[0])
        let pairs = [[a, "a"],[b, "b"],[c, "c"]]
        for (let pair of pairs){
            if (pair[0] > 0){
                maxHeap.enqueue(pair)
            }
        }
        let res = []
        while (!maxHeap.isEmpty()){
            let [count, char] = maxHeap.dequeue()
            // if the last two characters are the same as the one that was just popped out, we can't use this character to avoid making three in a row, so we pop out again to use the 2nd highest
            // count character, and then push the pair that we initially popped out back in again
            if (res.length > 1 && res[res.length - 1] === char && res[res.length-2] === char){
                // if there is no 2nd highest character, break
                if (maxHeap.isEmpty()){
                    break
                }
                // second highest count character
                const [count2, char2] = maxHeap.dequeue()
                res.push(char2)
                // push back in if the count is greater than 0 after using one char
                if (count2 - 1 > 0){
                    maxHeap.enqueue([count2 - 1, char2])
                }
                // push the initial one back in 
                maxHeap.enqueue([count, char])
            }
            else {
                res.push(char)
                if (count - 1 > 0){
                    maxHeap.enqueue([count-1, char])
                }
            }
        }
        return res.join("")
    }
}
